### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by ricokahler** [ricokahler/sanity-codegen#5]
Hey all 👋

Thanks for taking interest in this project. This one is especially beneficial to be open source because we can gain stability through our many difference scenarios and potentially work towards some really cool features.

Bellow a quick floating list of items I'm considering to add prior to a stable release:

- [ ] **CLI Stability** — not a feature per se but important regardless. I made a lot of assumptions with the CLI and I'm not sure it'll work for everyone so the more setups we run it through, the better it'll get
- [ ] **Codegen Stability** — same as above. The more schemas we run it through, the better
- [ ] **Support for additional types** — See (issue coming soon)
- [ ] **Support weak references** — Some references are weak and thus when expanding them, they should be defined as `T | null`
- [ ] **API codegen docs** — currently the API usage is relatively undocumented
- [ ] **GROQ projection codegen** — Similar to the some GraphQL libs, we should add support for parsing groq, finding what projection the users has written and pick and join keys from the existing types. (big effort)

Feel free to comment with wishlist features and don't hesitate to send any other feedback or bugs!

**Comment by PaulPCIO** [comment]
> API codegen docs — currently the API usage is relatively undocumented

To simplify the documents setup you can maybe opt for something like [nextra](https://nextra.vercel.app/), It is a really simple setup, and it supports md, mdx out of the box. 

I tested it today for a new project, and had some documents setup in a matter of minutes.

**Comment by OllieJT** [comment]
Hey, how's it going @ricokahler ?

I was about to make a PR to improve types for SanityImage but didn't know the status of this release.

**Comment by ricokahler** [comment]
@OllieJT feel free to PR improvements to the current version!

this release is still in an alpha state and won't be production-ready until a month or so (though I am planning on releasing a beta in a few weeks)

**Comment by ricokahler** [comment]
> I am planning on releasing a beta in a few weeks

So an update on this, I got overconfident on the previous approach and near the last 20%, it didn't pan out 😅.

TL;DR I'm rewriting the GROQ code generator from the ground up. See #149

**Comment by rix1** [comment]
Hi @ricokahler!
Not sure if I've come to the right place, but I'm testing out the alpha release and I noticed that the asset `metadata`  properties was fixed in https://github.com/ricokahler/sanity-codegen/pull/166 (thanks!). However, I don't think the the corresponding `@sanity-codegen/types` package was updated? Is that correct?

Is that depending on the rest of the 1.0 work, or would it be as simple as rebuilding and releasing the package? Not super familiar with TS, but let me know if I can help out in any way :v:

Btw, awesome project! 🙌

**Comment by rennehir** [comment]
Hi @ricokahler!

There has been almost a year since last commit. I'm pondering over using this library for my company and would like to know if you still have the intention to continue building this library?

**Comment by saiichihashimoto** [comment]
![image](https://user-images.githubusercontent.com/2819256/211632741-7904ab79-ceb5-4d0a-87c6-5f783fad96c1.png)

I think most of the work is going in the renovating, so I wouldn't look at `main` to see activity. Look like the last activity was end of November, which makes sense with holidays. I'd give this more time before deeming it stale. 😊

Another solution is [sanity-typed-schema-builder](https://github.com/saiichihashimoto/sanity-typed-schema-builder), which uses a builder and derived types to accomplish the same thing that codegen is accomplishing here. Shameless plug. To be fair, I think with the new `defineType` in Sanity v3, there's a good chance that derived types might be able to accomplish all of this.

**Comment by mosesoak** [comment]
Hi @saiichihashimoto - thanks for the lead on a possible alternative! The lag on codegen supporting v3 has made us drag our heels on upgrading our projects. 

> To be fair, I think with the new `defineType` in Sanity v3, there's a good chance that derived types might be able to accomplish all of this.

Can you elaborate on this? What we want codegen for is not the studio itself, since V3 should have us covered there, it's copying the generated .d.ts file into the frontend apps that connect to sanity -- that is the true benefit of generating types.

**Comment by saiichihashimoto** [comment]
I was hoping I could use typescript itself to derive the document types from `defineType` to use in the apps themselves, to know the type of the docs. I don't remember which repo I commented on that with, but I tried to accomplish this and ultimately didn't see how it was possible. This project and `sanity-typed-schema-builder` seem to truly be the best (and honestly, only) way to get typescript types for your documents in your app itself. I prefer `sanity-typed-schema-builder` for a few reasons (I have lots of beef with codegen in most cases as a concept), but ultimately there's pros and cons to both approaches. The reality is that it would be best if sanity made all of this irrelevant by giving us a solid first-class solution to this, which is likely doable through stronger typing of `defineType`.

So:
- short term: use `sanity-typed-schema-builder` or this project.
- long term: find the dead ends from deriving document type out of `defineType` and make PRs against sanity to type it stronger to allow this.

**Comment by genox** [comment]
Yes this seems to be an issue of weighing pros and cons, more than anything. It's up to us to decide which compromise we are willing to take. The ideal solution would be to have sanity.io treat TS document typing on the receiving end as a first class issue that should be resolved.

With V3, I prefer to use the "sanity way" of defining schemas as I assume (or hope..) that sanity will have us covered there in the future, somehow.

For the time being, I can use the old version of sanity-codegen just fine with a few tweaks here and there, and it is much, much better than nothing. Or much better than using another completely different approach to generate schemas first in order to achieve typed documents, I would really prefer to stay away from this, to keep the studio parts as vanilla as possible. Maybe we should start a petition to request first class ts support from sanity directly. ;-)

Maybe the maintainer of this project knows more about the future of sanity's TS plans than we do, because it doesn't make sense to invest time in to a feature that will be dropped on us half a year later from sanity eventually. But also, maybe that's just my wishful thinking..
